File: WidgetStyleProvider.swift
import SwiftUI
import WidgetKit

// MARK: - Widget Style Provider

struct WidgetStyleProvider {
    let size: WidgetSize

    // MARK: - Font Sizing

    func fontForSize(style: Font.TextStyle) -> Font {
        switch size {
        case .small:
            return smallFont(for: style)
        case .medium:
            return mediumFont(for: style)
        }
    }

    private func smallFont(for style: Font.TextStyle) -> Font {
        switch style {
        case .title:
            return .system(size: 16, weight: .bold)
        case .title2:
            return .system(size: 14, weight: .bold)
        case .headline:
            return .system(size: 13, weight: .semibold)
        case .body:
            return .system(size: 12, weight: .regular)
        case .caption:
            return .system(size: 10, weight: .medium)
        case .caption2:
            return .system(size: 9, weight: .regular)
        default:
            return .system(size: 12, weight: .regular)
        }
    }

    private func mediumFont(for style: Font.TextStyle) -> Font {
        switch style {
        case .title:
            return .system(size: 20, weight: .bold)
        case .title2:
            return .system(size: 18, weight: .bold)
        case .headline:
            return .system(size: 15, weight: .semibold)
        case .body:
            return .system(size: 14, weight: .regular)
        case .caption:
            return .system(size: 12, weight: .medium)
        case .caption2:
            return .system(size: 10, weight: .regular)
        default:
            return .system(size: 14, weight: .regular)
        }
    }

    // MARK: - Spacing

    var contentPadding: CGFloat {
        switch size {
        case .small: return 10
        case .medium: return 12
        }
    }

    var itemSpacing: CGFloat {
        switch size {
        case .small: return 4
        case .medium: return 8
        }
    }

    var sectionSpacing: CGFloat {
        switch size {
        case .small: return 6
        case .medium: return 16
        }
    }

    // MARK: - Progress Bar

    var progressBarHeight: CGFloat {
        switch size {
        case .small: return 4
        case .medium: return 6
        }
    }

    var progressBarCornerRadius: CGFloat {
        switch size {
        case .small: return 2
        case .medium: return 3
        }
    }

    // MARK: - Colors

    static func colorsForTheme(_ theme: ColorTheme) -> ProgressColors {
        switch theme {
        case .light:
            return ProgressColors(
                low: Color(red: 0.2, green: 0.7, blue: 0.3),
                medium: Color(red: 0.9, green: 0.8, blue: 0.1),
                high: Color(red: 0.95, green: 0.6, blue: 0.1),
                critical: Color(red: 0.9, green: 0.2, blue: 0.2)
            )
        case .dark:
            return ProgressColors(
                low: Color(red: 0.3, green: 0.85, blue: 0.4),
                medium: Color(red: 1.0, green: 0.9, blue: 0.3),
                high: Color(red: 1.0, green: 0.7, blue: 0.2),
                critical: Color(red: 1.0, green: 0.3, blue: 0.3)
            )
        case .auto:
            return ProgressColors(
                low: .green,
                medium: .yellow,
                high: .orange,
                critical: .red
            )
        }
    }

    static var progressBarBackgroundColor: Color {
        Color.gray.opacity(0.2)
    }
}

// MARK: - Color Theme

enum ColorTheme: String {
    case light = "light"
    case dark = "dark"
    case auto = "auto"
}

File: src/widget/WidgetStyleProvider.ts
import { ProgressColors } from '../types/ui';
import { WidgetSize } from '../types/widget';

export { ProgressColors, WidgetSize };

// MARK: - ColorTheme

export enum ColorTheme {
  Light = 'light',
  Dark = 'dark',
  Auto = 'auto',
}

// MARK: - Font Text Style

export enum FontTextStyle {
  LargeTitle = 'largeTitle',
  Title = 'title',
  Title2 = 'title2',
  Title3 = 'title3',
  Headline = 'headline',
  Body = 'body',
  Callout = 'callout',
  Subheadline = 'subheadline',
  Footnote = 'footnote',
  Caption = 'caption',
  Caption2 = 'caption2',
}

// MARK: - Font Descriptor

export interface FontDescriptor {
  size: number;
  weight: 'regular' | 'semibold' | 'bold';
}

// MARK: - Widget Spacing

export interface WidgetSpacing {
  padding: number;
  itemSpacing: number;
  sectionSpacing: number;
  progressBarPadding: number;
}

// MARK: - Font size maps

const SMALL_FONT_MAP: Record<FontTextStyle, FontDescriptor> = {
  [FontTextStyle.LargeTitle]: { size: 22, weight: 'bold' },
  [FontTextStyle.Title]: { size: 17, weight: 'bold' },
  [FontTextStyle.Title2]: { size: 15, weight: 'semibold' },
  [FontTextStyle.Title3]: { size: 13, weight: 'semibold' },
  [FontTextStyle.Headline]: { size: 12, weight: 'semibold' },
  [FontTextStyle.Body]: { size: 11, weight: 'regular' },
  [FontTextStyle.Callout]: { size: 10, weight: 'regular' },
  [FontTextStyle.Subheadline]: { size: 10, weight: 'regular' },
  [FontTextStyle.Footnote]: { size: 9, weight: 'regular' },
  [FontTextStyle.Caption]: { size: 9, weight: 'regular' },
  [FontTextStyle.Caption2]: { size: 8, weight: 'regular' },
};

const MEDIUM_FONT_MAP: Record<FontTextStyle, FontDescriptor> = {
  [FontTextStyle.LargeTitle]: { size: 28, weight: 'bold' },
  [FontTextStyle.Title]: { size: 22, weight: 'bold' },
  [FontTextStyle.Title2]: { size: 18, weight: 'semibold' },
  [FontTextStyle.Title3]: { size: 16, weight: 'semibold' },
  [FontTextStyle.Headline]: { size: 14, weight: 'semibold' },
  [FontTextStyle.Body]: { size: 14, weight: 'regular' },
  [FontTextStyle.Callout]: { size: 13, weight: 'regular' },
  [FontTextStyle.Subheadline]: { size: 12, weight: 'regular' },
  [FontTextStyle.Footnote]: { size: 11, weight: 'regular' },
  [FontTextStyle.Caption]: { size: 10, weight: 'regular' },
  [FontTextStyle.Caption2]: { size: 9, weight: 'regular' },
};

// MARK: - WidgetStyleProvider

export class WidgetStyleProvider {
  /**
   * Returns a size-appropriate font descriptor for the given widget size and text style.
   * Small widgets use reduced font sizes; medium widgets use standard sizes.
   */
  static fontForSize(size: WidgetSize, style: FontTextStyle): FontDescriptor {
    switch (size) {
      case WidgetSize.SMALL:
        return SMALL_FONT_MAP[style];
      case WidgetSize.MEDIUM:
        return MEDIUM_FONT_MAP[style];
    }
  }

  /**
   * Returns size-appropriate spacing values for the given widget size.
   */
  static spacingForSize(size: WidgetSize): WidgetSpacing {
    switch (size) {
      case WidgetSize.SMALL:
        return {
          padding: 10,
          itemSpacing: 4,
          sectionSpacing: 8,
          progressBarPadding: 2,
        };
      case WidgetSize.MEDIUM:
        return {
          padding: 14,
          itemSpacing: 6,
          sectionSpacing: 12,
          progressBarPadding: 4,
        };
    }
  }

  /**
   * Returns the appropriate progress bar height for the given widget size.
   */
  static progressBarHeight(size: WidgetSize): number {
    switch (size) {
      case WidgetSize.SMALL:
        return 6;
      case WidgetSize.MEDIUM:
        return 10;
    }
  }

  /**
   * Returns the appropriate progress bar corner radius for the given widget size.
   */
  static progressBarCornerRadius(size: WidgetSize): number {
    switch (size) {
      case WidgetSize.SMALL:
        return 3;
      case WidgetSize.MEDIUM:
        return 5;
    }
  }

  /**
   * Returns ProgressColors appropriate for the given color theme.
   * Colors meet WCAG AA contrast requirements.
   */
  static colorsForTheme(theme: ColorTheme): ProgressColors {
    switch (theme) {
      case ColorTheme.Light:
        return {
          low: '#33A652',
          medium: '#E69E00',
          high: '#DB3545',
          background: '#E5E5E5',
        };
      case ColorTheme.Dark:
        return {
          low: '#4DD973',
          medium: '#FFC226',
          high: '#FF6161',
          background: '#404040',
        };
      case ColorTheme.Auto:
        return {
          low: '#34C759',
          medium: '#FF9500',
          high: '#FF3B30',
          background: '#E5E5EA',
        };
    }
  }
}
